front/lib/llm/providers/mistral.ts

```diff
@@ -0,0 +1,240 @@
+    this.thinkingAccumulator += delta;
+    return {
+      type: "reasoning_delta",
+      content: {
+        delta: delta,
+      },
+      metadata: this.metadata,
+    };
+  }
+
+  // returns token usage and success/error completion events
+  private async *yieldFinalEvents(
+    event: CompletionEvent
+  ): AsyncGenerator<LLMEvent> {
+    const tokenUsage = event.data.usage;
+    if (tokenUsage) {
+      yield {
+        type: "token_usage",
+        content: {
+          inputTokens: tokenUsage.promptTokens ?? 0,
+          outputTokens: tokenUsage.completionTokens ?? 0,
+          totalTokens: tokenUsage.totalTokens ?? 0,
+        },
+        metadata: this.metadata,
+      };
+    }
+    const finishReason = event.data.choices[0].finishReason;
+    if (finishReason === "error" || finishReason === "length") {
+      yield {
+        type: "error",
+        content: {
+          code: finishReason,
+          message: finishReason,
+        },
+        metadata: this.metadata,
+      };
+    } else {
+      yield {
+        type: "success",
+        content: [
+          {
+            type: "text_generated",
+            content: {
+              text: this.textAccumulator,
+            },
+            metadata: this.metadata,
+          },
+        ],
+        metadata: {
+          ...this.metadata,
+          id: event.data.id,
+          created: event.data.created,
+        },
+      };
+    }
+  }
+
+  async *modelStream({
+    conversation,
+    prompt,
+  }: {
+    conversation: ModelConversationTypeMultiActions;
+    prompt: string;
+  }): AsyncGenerator<any> {
+    const messages = this.getMessages({
+      conversation,
+      prompt,
+    });
+
+    const events = await this.client.chat.stream({
+      model: this.model.modelId,
+      messages,
+      temperature: this.temperature,
+      stream: true,
+    });
+
+    for await (const event of events) {
+      yield event;
+    }
+  }
+
+  private async *streamMistralContent(
+    content: ContentChunk[]
+  ): AsyncGenerator<LLMEvent> {
+    for (const contentItem of content) {
+      switch (contentItem.type) {
+        case "text":
+          yield this.textDelta(contentItem.text);
+          break;
+        case "thinking":
+          for (const thinkingItem of contentItem.thinking) {
+            if (thinkingItem.type === "text") {
+              yield this.reasoningDelta(thinkingItem.text);
+            }
+          }
+          break;
+        default:
+          continue;
+      }
+    }
+  }
+
+  private getMessages({
+    conversation,
+    prompt,
+  }: {
+    conversation: ModelConversationTypeMultiActions;
+    prompt: string;
+  }): MistralMessage[] {
+    const messages: MistralMessage[] = conversation.messages
+      .filter((message) => message.role !== "content_fragment") // content_fragments should have been removed by now
+      .map((message) => {
+        let content: string | ContentChunk[];
+        if (typeof message.content === "string") {
```

Comment:
Let's make the condition on the role of the conversation message for more clarity. This is a good opportunity to understand the purpose of each message properties (e.g. `contents` vs. `content`)